### PR TITLE
Expand SeaORM entity generator with Seaography related data

### PR DESCRIPTION
**Pull request by karatakis**
## PR Info

- Related https://github.com/SeaQL/seaography/issues/130

## New Features

- [x] Add `seaography` flag to `sea-orm-cli`
- [X] Add generation of `seaography` related information to `sea-orm-codegen`
- [X] Add `DeriveEntityRelated` macro
- [X] Add generation of code that uses `DeriveEntityRelated` to `sea-orm-codegen`

### CODE REVIEW

File: sea-orm-codegen/tests/compact/cake.rs [hunk 1]
```diff
@@ -17,19 +17,12 @@ pub enum Relation {
...
-    }
-    fn via() -> Option<RelationDef> {
-        Some(super::cake_filling::Relation::Cake.def().rev())
-    }
+#[derive(Copy, Clone, Debug, EnumIter, DeriveRelatedEntity)]
+pub enum RelatedEntity {
+    #[sea_orm(entity = "super::fruit::Entity", to = "Relation::Fruit.def()")]
+    Fruit,
+    #[sea_orm (entity = "super::filling::Entity", to = "super::cake_filling::Relation::Filling.def()", via = "Some(super::cake_filling::Relation::Cake.def().rev())")]
+    Filling
```
Comment (billy1624) [inline sea-orm-codegen/tests/compact/cake.rs:25]: I think we should keep generating the plain old `Related<T>`. With that in mind, if user pass in a `--seaography` flag to the `sea-orm-cli generate entity`, then, an additional `RelatedEntity` enum will be generated.

I imagine the derive attribute of each variant in the `RelatedEntity` could be... to take advantage of the existing `Related<T>` impl and avoid saying the same thing twice.

```rs
#[derive(Copy, Clone, Debug, EnumIter, DeriveRelatedEntity)]
pub enum RelatedEntity {
    #[sea_orm (entity = "super::filling::Entity")]
    Filling
}
```

File: sea-orm-codegen/tests/with_seaography/cake.rs [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+impl seaography::RelationBuilder for RelatedEntity {
+    fn get_relation(&self, context: & 'static seaography::BuilderContext) -> async_graphql::dynamic::Field {
+        let builder = seaography::EntityObjectViaRelationBuilder { context };
+        match self {
+            Self::Fruit => builder.get_relation:: <Entity, super::fruit::Entity>("fruit"),
+            Self::Filling => builder.get_relation:: <Entity, super::filling::Entity>("filling"),
+            _ => panic!("No relations for this entity")
+        }
+    }
+}
```
Comment (billy1624) [inline sea-orm-codegen/tests/with_seaography/cake.rs:49]: These `async-graphql` constructors for relations can be moved into `sea-orm-macros` in which those methods will be generated when the `seaography` feature is enabled

**Review by billy1624 — COMMENTED** [review]
Hey @karatakis, thanks for the huge PR! Great work! I just leave a few comments :)

**Comment by billy1624** [comment]
Hey @karatakis, thanks for the changes!! Please feel free to ping me once it's finished or you need help :P

**Comment by karatakis** [comment]
I am done.

@billy1624 could you please take a look if the last changes resolve the issues you pointed out ?

**Comment by karatakis** [comment]
I used the proposed changes in the following https://github.com/SeaQL/seaography/pull/131
Its still WIP but you can see how I am planning to use it :)

**Comment by tyt2y3** [comment]
Great work. This is really significant. We will try to roll out SeaORM 0.12 sooner and I think this is really a 'killer app'

**Review by billy1624 — COMMENTED** [review]
Hey @karatakis, look good!! Neat and clean!

Can we update the `Cargo.toml` in https://github.com/SeaQL/seaography/pull/131 and make it depends on this PR. I want to compile, see it in action and play with it :)

**Review by tyt2y3 — APPROVED** [review]
I think we can schedule a merge @billy1624 if there is no other outstanding PR to sea-orm-cli
We want to be extra careful not to breaking any existing features though.

**Review by billy1624 — COMMENTED** [review]
Hey @karatakis, I just added two commits

1. Derive `impl seaography::RelationBuilder for RelatedEntity` when `seaography` feature is enabled https://github.com/SeaQL/sea-orm/pull/1599/commits/0833478a79e1d5c4bbbed89923d3af2d61d90e5c

2. A demo for how to construct the `async-graphql` query root for entity with/without `RelatedEntity` https://github.com/SeaQL/sea-orm/pull/1599/commits/c84294edb3a9bd302ffed97a5294d266e58fbc3e

Please check :)

**Comment by tyt2y3** [comment]
The seaography feature in the Cargo.toml needs a '?'

File: Cargo.toml [hunk 3]
```diff
@@ -122,3 +122,4 @@ runtime-tokio-rustls = [
     "runtime-tokio",
 ]
 tests-cfg = ["serde/derive"]
+seaography = ["sea-orm-macros/seaography"]
```
Comment (tyt2y3) [inline Cargo.toml:125]:
```suggestion
seaography = ["sea-orm-macros?/seaography"]
```

**Comment by billy1624** [comment]
Indeed! Thanks @tyt2y3 for catching that

**Comment by billy1624** [comment]
Emmmm... no. We're wrong. `sea-orm-macros` is not a optional dependency. It's always on, just that we pass-through `sea-orm-macros`'s feature via `sea-orm`'s feature.

**Review by tyt2y3 — APPROVED** [review]
No questions then!

**Pull request merged**